spec/producer_spec.lua

```diff
@@ -231,19 +233,72 @@ describe("Test producers: ", function()
 
   it("buffer flush", function()
     local p = producer:new(broker_list_plain, { producer_type = "async", batch_num = 1, flush_time = 10000})
-    ngx.sleep(0.1) -- will have an immediately flush by timer_flush
+    ngx_sleep(0.1) -- will have an immediately flush by timer_flush
 
     local ok, err = p:send(TEST_TOPIC, nil, message)
     assert.is_not_nil(ok)
     assert.is_nil(err)
-    ngx.sleep(1)
+    ngx_sleep(1)
     local offset0 = p:offset()
     p:flush()
     local offset1 = p:offset()
     local offset_diff = tonumber(offset1) - tonumber(offset0)
     assert.is.equal(offset_diff, 0)
   end)
 
+  it("works when broker is down and brought back online (sync)", function()
+    local p, offset, ok, err
 
+    p, err = producer:new({ { host = "broker", port = 9092 } }, { socket_timeout = 15 * 1000, max_retry = 3, retry_backoff = 5 * 1000 })
+    assert.is_truthy(p)
+    assert.is_nil(err)
+    offset, err = p:send("brokerdown", key, "beforestop message")
+    assert.is_truthy(tonumber(offset) > 0)
+    assert.is_nil(err)
+
+    ok = os.execute(string.format("docker compose -p dev %s broker2", "stop"))
+    assert.is_truthy(ok)
+
+    ngx_sleep(5)
+    offset, err = p:send("brokerdown", key, "afterstop message")
+    assert.is_nil(offset)
+    assert.is_same("not found broker; not found broker; not found broker", err)
+
+    ok = os.execute(string.format("docker compose -p dev %s broker2", "start"))
+    assert.is_truthy(ok)
+
+    ngx_sleep(15)  -- wait for rediscovery
```

Comment:
Did you test for flakyness? Tests that are bound to timing tend to fail randomly.

Reply:
Tested. Most of the time, `1~5` seconds is enough, `15s` is a safeguard. I increase this value to make it robust.

Consider this repo is less updated than kong or kong-ee, I prefer not to increase the 'helpers.wait_until' utility.